Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -254,240 +168,272 @@ namespace Terrain
                 }
             }
         }
     }
 
-    void TerrainFeatureProcessor::InitializeTerrainPatch()
+    void TerrainFeatureProcessor::InitializeTerrainPatch(uint16_t gridSize, float gridSpacing, PatchData& patchdata)
     {
-        m_gridVertices.clear();
-        m_gridIndices.clear();
+        patchdata.m_positions.clear();
+        patchdata.m_uvs.clear();
+        patchdata.m_indices.clear();
+
+        uint16_t gridVertices = gridSize + 1; // For m_gridSize quads, (m_gridSize + 1) vertices are needed.
+        size_t size = gridVertices * gridVertices;
+        size *= size;
 
-        for (float y = 0.0f; y < m_gridMeters; y += m_gridSpacing)
+        patchdata.m_positions.reserve(size);
+        patchdata.m_uvs.reserve(size);
+
+        for (uint16_t y = 0; y < gridVertices; ++y)
         {
-            for (float x = 0.0f; x < m_gridMeters; x += m_gridSpacing)
+            for (uint16_t x = 0; x < gridVertices; ++x)
             {
-                float x0 = x;
-                float x1 = x + m_gridSpacing;
-                float y0 = y;
-                float y1 = y + m_gridSpacing;
-
-                uint16_t startIndex = (uint16_t)(m_gridVertices.size());
-
-                m_gridVertices.emplace_back(x0, y0, x0 / m_gridMeters, y0 / m_gridMeters);
-                m_gridVertices.emplace_back(x1, y0, x1 / m_gridMeters, y0 / m_gridMeters);
-                m_gridVertices.emplace_back(x0, y1, x0 / m_gridMeters, y1 / m_gridMeters);
-                m_gridVertices.emplace_back(x1, y1, x1 / m_gridMeters, y1 / m_gridMeters);
-
-                m_gridIndices.emplace_back(startIndex);
-                m_gridIndices.emplace_back(aznumeric_cast<uint16_t>(startIndex + 1));
-                m_gridIndices.emplace_back(aznumeric_cast<uint16_t>(startIndex + 2));
-                m_gridIndices.emplace_back(aznumeric_cast<uint16_t>(startIndex + 1));
-                m_gridIndices.emplace_back(aznumeric_cast<uint16_t>(startIndex + 3));
-                m_gridIndices.emplace_back(aznumeric_cast<uint16_t>(startIndex + 2));
+                patchdata.m_positions.push_back({ aznumeric_cast<float>(x) * gridSpacing, aznumeric_cast<float>(y) * gridSpacing });
+                patchdata.m_uvs.push_back({ aznumeric_cast<float>(x) / gridSize, aznumeric_cast<float>(y) / gridSize });
             }
         }
-    }
 
-    bool TerrainFeatureProcessor::InitializeRenderBuffers()
+        patchdata.m_indices.reserve(gridSize * gridSize * 6); // total number of quads, 2 triangles with 6 indices per quad.
+        
+        for (uint16_t y = 0; y < gridSize; ++y)
+        {
+            for (uint16_t x = 0; x < gridSize; ++x)
+            {
+                uint16_t topLeft = y * gridVertices + x;
+                uint16_t topRight = topLeft + 1;
+                uint16_t bottomLeft = (y + 1) * gridVertices + x;
+                uint16_t bottomRight = bottomLeft + 1;
```

Comment:
While simple and correct, this is not going to leave you with a very cache friendly vertex buffer unless gridVertices is small. If this is going to stick around for a while, consider creating a follow up task to add a step that does some basic vertex cache optimization.

Reply:
That's a good callout for the future and something I should chat with you about. Right now things are always changing so it's good to keep it simple. We'll probably be changing the triangulation pattern, and may even just build the meshes with heights on the c++ side since ray tracing will need that (at least until ray tracing can do height intersections or we have a way to generate meshes on the GPU that ray tracing can use)